src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogValuesFormatter.cs

```diff
@@ -205,7 +204,19 @@ private object FormatArgument(object? value)
             var enumerable = value as IEnumerable;
             if (enumerable != null)
             {
-                return string.Join(", ", enumerable.Cast<object>().Select(o => o ?? NullValue));
+                var vsb = new ValueStringBuilder(stackalloc char[256]);
+                bool first = true;
+                foreach (object e in enumerable)
+                {
+                    if (!first)
+                    {
+                        vsb.Append(", ");
+                    }
+
+                    vsb.Append(e != null ? e.ToString() : NullValue);
+                    first = false;
+                }
```

Comment:
What you have is fine.  An alternative way to write it would be:
```C#
string separator = "";
foreach (object e in enumerable)
{
    vsb.Append(separator);
    vsb.Append(e != null ? e.ToString() : NullValue);
    separator = ", ";
}
```
No strong preference about which way is used.

Reply:
I guess using the `first` bool is just the way I think about it. I put it in my benchmark, and it looks like the bool approach is just barely faster:

```C#
        [Benchmark]
        public string NewWay()
        {
            var vsb = new ValueStringBuilder(stackalloc char[256]);
            bool first = true;
            foreach (object e in s_Enumerable)
            {
                if (!first)
                {
                    vsb.Append(", ");
                }

                vsb.Append(e != null ? e.ToString() : NullValue);
                first = false;
            }
            return vsb.ToString();
        }

        [Benchmark]
        public string StevesWay()
        {
            var vsb = new ValueStringBuilder(stackalloc char[256]);
            string separator = null;
            foreach (object e in s_Enumerable)
            {
                vsb.Append(separator);
                vsb.Append(e != null ? e.ToString() : NullValue);
                separator = ", ";
            }
            return vsb.ToString();
        }
```

|    Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
|    OldWay | 493.9 ns | 1.24 ns | 1.03 ns | 0.0794 |     - |     - |     504 B |
|    NewWay | 215.2 ns | 0.54 ns | 0.42 ns | 0.0621 |     - |     - |     392 B |
| StevesWay | 220.5 ns | 1.21 ns | 1.13 ns | 0.0617 |     - |     - |     392 B |

I compared [the assembly](https://sharplab.io/#v2:C4LghgzgtgPgAgBgARwIwDoBCBXAZrgUwCcIBuAWAChEUMARASzAHMA7AewmAYGMyqaadACVsrblALoAwuygAHBgBtiAZWIA3XgX7VkQ0eIaT0ASXHF289US08dFPbXQAVAgA9gjwRllKVPNzsrLoCAMwoAExI0kgA3lRISckpSfJEDBpgwARIPMFctMgActj+AGpgSti5ALxIAEQAFKxlSgCUDY6JqUlwEWglBADuAOpgAJ5NaBEErZJE7T29CZS9vVlESBoQAEZI9awjSJXVBKrAGazMOMoAJsRNXGA8ANZVSuw8eQAWYEQAbUiAFYAGwAXXajnWqV27HYSiQuAYJGAByQlxq0JhyVw7CIBBePyQTXYuwAVgRAkhcgxWDT5sQlmscUlVqzegxcCSAITI1HMjmpdlC9Y7XboACC8nkczuTQaABpGlDlqKAL5UNVC8VSmVypq5HmHNpIAD8NNc7AuV2YTXaSBASFKFSqNVVLNFSJRhXquCqEAI2NZms9OLgAHZtnsrTa6XaPTDQ9rkv0ikgLgQNDpxlMZgzsAtBesRazNtH9odjqcanHrrclA8iE9gC93v4vr9/kCwZDg+HUMhA/J/tl8ejWv5+zC8QSiSSyZTqbT6XNC0yUytNzjddLZax5cPR8B8YmvRW9fv5UaTf5zZaXNbLvH7Y7nW0awQz16j0Qx1t6iVRpp3WUMhUjC9HzrBMQKSUNk0oAB6RCkAAGW0EICDuDF2AxH5cnQYoAFEXCQAAxdgxDubIGGCJAqOIJBglyccoHxXIWAJAhJHECB0CoZCkBcfCkEIkjyMog8aLopQMMDCA8IYBTkRUHCkAmSj6IPRjgBEgBZUxSNk+xMP4ygtUoVgwEkCAR3sZw3E8KhSyQOkciIKzEQJbkRyIbgqiQLgiGwalP2ghsmzVFyUnSTJslyHg/kBcELQAfX+P8JkfYQCGAbAPMfAAFeElFgpBYqyHIMxHVgAB5Ev+AA+JBUoakgyoq+LXPEFr5E4bowxSNMwufetsHuR5VBq+qkuaukGH8pQcHwDdBuSaL1nSohMuy3L8tYIqSonNoypSVqkoU+p5sW5bCCIU7klSvrLqQBAyoQmFhrdc5RpucbG0eNzuoWpglGkMA7IWiZiy3Nbei2nb2ByvKCvYYqEXRSVtsmdGlBmpr0FUJKsJEOZgCaa7QfByHgGhh6knO/4XoRyZdpRg60ZK+nes4dE3u3D71jTIHULmZhdO3DbemYXKDmap7+u3FJA2AJX1rV3o6AIXZsGYKUIEDPymiyM4kEa+oEG/UUtZ1vXJQN4hyZNmokFq+pGZIdBReuXSraFBWXudoMNaQMDQIsz6IiB6mXihuWWravjvfFn4BsjlAABYkCIkJ8oIGOeChimep4CHY9pmHhRDwThKU1yFI4NEPFlQIsLU3YEo+NvdLr4YFuJI5mBo7M8jLwvaZDm3df1w3ydLmmJjNi2PWrlDTG5XTO/8Ri/gUsAkFtpAwAPUf/AUnuFL73SkEH4fO4X5UoDAV5cggPO1NLu8AHEiHYYY1N0r/f+x8aTuHsPIIIrAzJCi5CSJo41xDtHnuXRezV4FuXaB7JOYtfYhxSD/P+TRkHj0XgAWh5hAP2yRBa9GQihWqb8oBPyIBMRq246FIC/rLfeihWBWV2KpbyxA5j2RPHhXIOsJpEGgesDhdB2A6BvuwJuucCTiLUXXfek5ERtSPrgdyrtAwJW8rUBoyddINEQmwuGSAOG1wUuwbMRBPhgGwnXEcwB3L0ifsARKbc6TiJiAAYiQBGdAYQADUeQ5CKBUFsCAuENLYFHvSOQC12GCU3qA+QxkFpIEkLpdg2FP5KGVMfbCwwMhVQIMwRoyJ3BtyIUlAAVHkdEkiAaLAaBk+hiFGHMNYduNM3kuxbC4cAQqdJ+EqByndERBB7SSxDhBEZeluL4gmHpJmfwlDoHGbM4RrB7BNCwVQuCEdZGCQYYWAZ1iYQcPGUfcqUywACNyEIgkRzchiKyR0psMjaFXL6Tc/4gybEcNqr5ayN9rIEFMe5KAdJ4oNEajnN+BJz5/DRL8/6TYkC7yedo0ZeiDEMIIMYgguBTHmJ+JYxqtVEJQqgHcoWEQRm6PGZMvhryZmUsOccuEGMEVIpyJXFIUtUiwKaMKqyoqQ4SphGivOBci40qQFE1AZycRYIBDS8E6IADkAAdBABruZh1ZCsylSA1lsRYVskgOy9m5QOZ845pz3oXN6MM61uiL4AiBnSB47hwRLJsSkGWqtw3q2jZrbW097az2LsG12FCtWfSjCMnVQaPDgnNQLL1qQ0yOOIBkB46YoK/UWTYhVqZByBXRFgwmxkFkIGVAHdosYq3puSIwWynAFk9r6FGXQSZC0pAhf00FjVkb7UxbkBiziJjxkCieP8MsmIbxEn84g6AGVTpYSy71EQprH3xkQZqwgwDDGkBdeOWC06XPoQesF9zBKzo8nvQKNUj6/yooE/IFheKbsCTu6RPTXbAqYdOiDkLRxQBhZIeFxBEWyoICipVGK8LZFA7inekBCWml0WAfRjEyUUqpWYnBtKrEMqZUeotEQcquIAPKsCUBMU9dU2rNXtlxpogrEQyvimKmNMDuTSpQyKr88q8HJEw/nMeqrqPqqQJq7mZ1E66uo/q+oxrTX5tjSgKMTbVAtqaG2tNnrzI2LTMxu4bGONcfPbxiA/GHS1HlonZt2gLPtuelquzhIHPsc49NHjSA+M1WLmiZ4fkPNeYuj545cXgD+d5uQ1LgWmPBcc2Fs9EWovHxi6u/4aXupohUD7H4CWE5JbM751LyoqspxXrZiIgmhIsNkPILK7AmjOYiw8LgIraKsGVJRNEQNE6jCqTkVgom2QhylaZ8zlmO2uG61YPrTRhvcFlWN9oi2q5GZSDNubcxG3PQ072pSz1B03eHRiYKwcjMWpxAQJQgZZOneSOdha82+aPaQH2+79pgcQX9F917Qp3spBoYxzOSBzBJsDdpdwypdFB2VNNySwBju1qSCt56Zs6tMy9ip8h+QxD45++eThQCiF46HaHcd6wgYEifvNa4V2MvdWDdzVbvmc0Y6QJzsA3PmCdp6ztoXxyReqep4glncuFki8x8z9AZFlBKGNt9FXJOIn1CV14AtNn05wCzijx2JX1dFAtJQsN4mSSXVvB0On55IyGeWyXPH6JsHVcfayYnvM0FNrVVT5nx3xVyaSAQ4YTOacs4RzCDn3EJesBXe7En5CReC+8w1+X6PlTi8l9L7bj4TkF/MwrqJJujvcz4kV1g9oZAV/66r5NHgG8hwDuq43ePrPbgBGswpdxTAKF16Pn4RSJ85JYxAsbTfmDMAxRATIBBzCyUz9cdoob2tI73AaXRPACc+7RCT7PisjNSvQYgkntUSQIPx+H6j0exPnh1c9XTeRgd9/qIbmpoZkKJ9t9kZoTqkPHpKAeEfgeEQsnmbsPtPrPpPk0MgePpPgvpAsvqvjoOvtmFvmXvvhblnLAfKIMA7mfjfhJq7jfG0O/ktr9sOt7jfj1JfhQtzFKgHinAcPUKgEgAAGQCFP4YIP4iGIKv7VYOiCROKLz5BMJ0SlyBjKhSB6xHz6gHgrqDDKS/wIalBQAdxEAUREA+LmB4hICyQvyBQEBQproQDKiyhEAmSxYTAGEIj2E0i+IAo4gQGaYXQAjf7+4AgIB5qx4ULoiAGoDAEcigEEAe5ehkFmaEKULRG9Ap7JCdRVSW6RYaF3BJEJ6DCBRUEwJsG8xX6jochSok6Jbk4R6BQU5SHxGijx5PANGtYsFChN5ubRbl69aV6d4BYG68xG71E0pD4H7ZFkFNL/B5A46+5J5O6VESZ941Gex1H15NFCgtH14dEciDazRIB3C+hk6eyF4LIBwa5J7cyzgkiBpA6uSppHHABtG6SkCuQRIRIMHxBhFPEBo/48CpHaqG4D407jHpxiAQAka5CTG5HTFECtLY4VYWFv6LHB6lHMzXbLYSbVEnHcHXzkIta4LgFhEtGEk1a7Gsj7FNSHHHGd4XHIlSHXHji273EMCPFcAvGpzvGfGbEci/EMA/7NJBwfGAkwh94jFklgmsqH6wn2Z5ZUkXrbD66ok4hAzsEBycHYmh64mclIDkJByclfG+HJAtEGk0oIHm4cgGky79HV6+YdpDEKQjFmnUZSm9Aj65Qz4YE5JoGekoHz6L4FBSgr5r4b6EE75S7EHSkKm8a5H8ZAxknFGVE9T4gMDMDFQYnX7O6kgZDpnamSE8EEkonElMEM6EKJkUlAllFMS5kZmqaSnLImZ2nHKpl5keEVmIE2Ielj5z5T5+nelKBYFL6ETsDhnxh77biZHQlZxQEwGwkn5JmsgtHqYhxTGn5umpCTogqHoQY5Tr4ABe3yIkbkxAnkB8eAd0NI/cjEuwi8dwlEAiK6PA+UnyaIOsK08SDAh5NZEGt5R8dwdwK6cGf4UAIyVkSGDQrigFkCVQKqtMmABAGkB4GZdKOEsGRieQJiDQAcqFwwPwvA+ETi9cSAuB8U4G4KQKL6DGE6VyTKiGcKkFAFIMwQsFSm8FiFwQdwKF1FyQHC+hhhIGicYuBAAAjjUFwG3B3EhcUi+WTOVJwMxVArBoyvBjxUkN2V6b2b6T2ZgYGSECOWObvlGb0FOUji0UDFBYpaxQvAhdJRmYuTiFPHbA7EbJZTBWDGxRMLZZxXWc1JbNzE5TPDbisTqXUW5WNtZSgt5chZwMqA0PHqfCoG4kcviASIEBxsqBwMJQebSA3OSg8HcOgA0G1kKIJFslYeit8rhCoGiDlD1IAn/EfKuGAjYZAq5FulvHEvigRvvIfOUoEsNiiG3MQnHHXLfNwNmCHG1ACPqn1AiFjJlJjNjBMLjC5oTMTIVXVeTDFu0FsrpOgFsu4E0HfvjicoAeFSxR5TZRxTFZQsqCdZgt5mqq0pED3paayJ3utgFm3n0f1nNUoAtZMCVRyNNSlDhB+vSO7BlKzEjHtKjLjPnnelDctWzPtIdBjAASVIDRMJqSSCeBDUgMaHQf4EaWEdjatTxutf8CTBDdKrDezBaWOu9akBpf6X2TpQGdgcGbgQbGGexkQUMgMFnKDgOtWjCBAaDRaPjXDZDS1NDX1hDejaVCHBfOiA8P6GUF4LYihDcZCYQLTMqGImANkTQP9W3PLe1YpApHSM8F8sRaWsxJRBABxuoVeN3LhCwBnliXjfTftITW7qTaWeTSVGtUTNTZtTLXTRDYzeHGGPBFQEAA=), and from what I can tell the loading of a variable to call `Append` is what makes it slightly slower.

![image](https://user-images.githubusercontent.com/8291187/106972733-0de5ba00-6717-11eb-929a-2dbe4f1a21b4.png)

So I kept it as is.

Reply:
Yup,  it wasn't a perf suggestion, just a conciseness one.